core/src/main/java/io/questdb/cairo/O3PurgeDiscoveryJob.java

```diff
@@ -28,181 +28,266 @@
-                                LOG.error()
-                                        .$("purge queue is full [table=").$(tableName)
-                                        .$(", ts=").$ts(partitionTimestamp)
-                                        .$(", txn=").$(nameTxnToRemove)
-                                        .$(", errno=").$(errno)
-                                        .$(']').$();
-                            }
-                        } else {
-                            LOG.error()
-                                    .$("could not purge [table=").$(tableName)
-                                    .$(", ts=").$ts(partitionTimestamp)
-                                    .$(", txn=").$(nameTxnToRemove)
-                                    .$(", errno=").$(errno)
-                                    .$(']').$();
-                        }
+        // find duplicate partitions
+        assert partitionList.size() % 2 == 0;
+        Vect.sort128BitAscInPlace(partitionList.getAddress(), partitionList.size() / 2);
+
+        long partitionTs = Numbers.LONG_NaN;
+        int lo = 0;
+        int n = (int) partitionList.size();
+
+        path.of(root).concat(tableName);
+        int tableRootLen = path.length();
+        try {
+            txnScoreboard.ofRO(path);
+            path.trimTo(tableRootLen);
+            txReader.ofRO(path, partitionBy);
+            loadLastTx(txReader);
+
+            for (int i = 0; i < n; i += 2) {
+                if (partitionList.get(i + 1) != partitionTs) {
+                    if (i > lo + 2) {
+                        processPartition(
+                                ff,
+                                path,
+                                tableRootLen,
+                                txReader,
+                                txnScoreboard,
+                                partitionTs,
+                                partitionBy,
+                                partitionList,
+                                lo,
+                                i
+                        );
                     }
+                    lo = i;
+                    partitionTs = partitionList.get(i + 1);
                 }
             }
-            return true;
+            // Tail
+            if (n > lo + 2) {
+                processPartition(
+                        ff,
+                        path,
+                        tableRootLen,
+                        txReader,
+                        txnScoreboard,
+                        partitionTs,
+                        partitionBy,
+                        partitionList,
+                        lo,
+                        n
+                );
+            }
+        } catch (CairoException ex) {
+            // It is possible that table is dropped while this async job was in the queue.
+            // so it can be not too bad. Log error and continue work on the queue
+            LOG.error().$(ex.getFlyweightMessage()).$();
```

Comment:
This does log OS error number afaik

Reply:
This logs
```
2022-01-12T10:09:53.372961Z E i.q.c.O3PurgeDiscoveryJob could not open read-write with clean allocation [file=/tmp/junit15477711821732187212/dbRoot/tbl/_txn_scoreboard]
```

Reply:
that's my point, "could not open" has to be supplemented with OS error code

Reply:
Updated in https://github.com/questdb/questdb/pull/1778/commits/4cd8ecd9b8d0647988fda61943d386c0a2d5ab46